qkeras/quantizers.py

```diff
@@ -440,6 +441,555 @@ def non_trainable_variables(self):
+           tf.equal(self.bits, K.cast_to_floatx(1)), 
+           lambda: K.cast_to_floatx(2), 
+           lambda: self.int_repr_max - self.int_repr_min
+        )
+
+    @tf.function
+    def __call__(self, x):
+        """Core quantization function"""
+        # build if not done so already
+        self._build()
+        shape = x.shape
+
+        # Data type conversion
+        x = K.cast_to_floatx(x)
+
+        # determine if using sign function
+        xq = tf.cond(
+            tf.equal(self.bits, K.cast_to_floatx(1)),
+            lambda: self._sign_function(x),
+            lambda: self._multi_bit_computation(x),
+        )
+
+        res = x + self.qnoise_factor * (xq - x)
+        # Annoying hack to deal with tf not knowing shape of output
+        res.set_shape(shape)
+
+        return res
+
+    def _multi_bit_computation(self, x):
+        """Quantization multi-bit representation- differs for auto and static alpha"""
+
+        xq = tf.cond(
+            self.auto_alpha,
+            lambda: self._auto_alpha_computation(x),
+            lambda: self._static_alpha_computation(x),
+        )
+
+        return xq
+
+    def _auto_alpha_computation(self, x):
+        """Compute quantized value for automatically determined alpha"""
+
+        # Get the minimum floating point scale that does not clip the max of x
+        alpha_scale = self._get_alpha_scale(x)
+
+        # Autoscale functions for tf.cond below
+        def autoscale():
+            """Quantize with `alpha_scale` above"""
+            int_xq = self._get_quantized_integer(x, alpha_scale)
+            return alpha_scale, int_xq * alpha_scale
+
+        def po2_autoscale():
+            """Get the "best" po2 scale for the data"""
+            _alpha_scale, int_xq = self._po2_autoscale(x, alpha_scale)
+            return _alpha_scale, int_xq * _alpha_scale
+
+        alpha_scale, xq = tf.case(
+            [
+                (tf.equal(self._alpha_str, tf.cast("auto", tf.string)), autoscale),
+                (tf.equal(self._alpha_str, tf.cast("auto_po2", tf.string)), po2_autoscale),
+            ],
+        )
+
+        # save scale for later computations
+        self.scale.assign(alpha_scale / self.integer_repr_scale)
+
+        return xq
+
+    def _sign_function(self, x):
+        """Sign indicator function for 1-bit quantization"""
+
+        neg_res = K.cast_to_floatx(-self.keep_negative)
+        nonneg_res = K.cast_to_floatx(1)
+
+        xq = tf.where(tf.math.greater_equal(x, 0), nonneg_res, neg_res)
+
+        return xq
+
+    def _get_alpha_for_sign_function(self, x):
+        """Set alpha scale for sign function"""
+
+        # Note: does not affect the actual quantization
+        alpha_scale = tf.case(
+          (tf.equal(self._alpha_str, ''), lambda: K.cast_to_floatx(1)),
+          (tf.equal(self._alpha_str, 'auto'), lambda: self._get_alpha_scale(x)),
+          (tf.equal(self._alpha_str, 'auto_po2'), lambda: self._po2_round(self._get_alpha_scale(x))),
```

Comment:
Why do a simple round here instead of the more advanced quantization error optimization for po2 scale?

Reply:
Totally redid this